### PR TITLE
cuDNN supports batch major data inputs under TF2 Keras RNN APIs

### PR DESCRIPTION
**Pull request by kaixih**
The CuDNNRNNV3 op is able to process the inputs with both time- and batch-major format.

Previous Keras RNN API will transpose the batch-major input to time-major before invoking the cudnn_rnnv3.

This PR skips the transpose and directly calls the cudnn_rnnv3.

fyi. @nluehr

### CODE REVIEW

**Review by qlzh727 — COMMENTED** [review]
Thanks for the change. It will be nice to have batch major input for unmasked data.

File: tensorflow/python/keras/layers/recurrent_v2.py [hunk 1]
```diff
@@ -486,9 +486,12 @@ def step(cell_inputs, cell_states):
 def cudnn_gru(inputs, init_h, kernel, recurrent_kernel, bias, mask, time_major,
               go_backwards):
   """GRU with CuDNN implementation which is only available for GPU."""
-  if not time_major:
+  if not time_major and mask is None:
     inputs = array_ops.transpose(inputs, perm=(1, 0, 2))
```
[the remark this answers is not in the document]
Comment (houtoms) [reply tensorflow/python/keras/layers/recurrent_v2.py]: This PR will skip the transpose when sequence_lengths/mask is provided (indicating the v3 API will be used and only v3 supports batch major).

[the remark this answers is not in the document]
Comment (qlzh727) [reply tensorflow/python/keras/layers/recurrent_v2.py]: Agreed, I wasn't notice this when I look at the start of the function. I can approve the PR as is, but I think it will be nice to have v3 enabled for non-masking case as well.

[the remark this answers is not in the document]
Comment (houtoms) [reply tensorflow/python/keras/layers/recurrent_v2.py]: Sure. Actually, the cudnn rnn ops in contrib has already done it by preparing the seq_len for users. https://github.com/tensorflow/tensorflow/blob/a232fda68d692da0f409608dd5960d63ac700e5d/tensorflow/contrib/cudnn_rnn/python/ops/cudnn_rnn_ops.py#L1133.

I can do another PR for this feature when this is done.

[the remark this answers is not in the document]
Comment (qlzh727) [reply tensorflow/python/keras/layers/recurrent_v2.py]: oh, I was wondering if this can be done on the c op level for performance reason. It will be nice If the implementation of cudnnrnn_v3 can take care of the absent of sequence length rather than force user to provide one.

[the remark this answers is not in the document]
Comment (houtoms) [reply tensorflow/python/keras/layers/recurrent_v2.py]: I think it could be done. But I don't see the performance issue for letting python prepare the seq_len, since this array is on host memory and will be on host memory even for cudnn calls.

[the remark this answers is not in the document]
Comment (qlzh727) [reply tensorflow/python/keras/layers/recurrent_v2.py]: I guess you have better knowledge for the cudnn kernel, I original idea was that for non-mask input, the generation of seq_len tensor is not needed at all (just process the whole input sequence).

[the remark this answers is not in the document]
Comment (houtoms) [reply tensorflow/python/keras/layers/recurrent_v2.py]: Yes, let me check how much work need to be done for generating seq_len inside the op (which might require the API change, because V3 requires seq_len).

But for this PR, I think it is ready to go.

File: tensorflow/python/keras/layers/recurrent_v2.py [hunk 2]
```diff
@@ -486,9 +486,12 @@ def step(cell_inputs, cell_states):
...
-  init_h = array_ops.expand_dims(init_h, axis=0)
+    seq_axis, batch_axis = (0, 1)
+  else:
+    seq_axis, batch_axis = (0, 1) if time_major else (1, 0)
+  init_h = array_ops.expand_dims(init_h, axis=seq_axis)
```
Comment (qlzh727) [inline tensorflow/python/keras/layers/recurrent_v2.py]: For the init_h here, it doesn't contain a sequence dim (it only contain one time step), should the expand_dim always apply to 0?

Comment (houtoms) [reply tensorflow/python/keras/layers/recurrent_v2.py]: No, it doesn't. The input_h shape is [num_layers, batch, num_units] for time_major, or [batch, num_layers, num_units] for batch_major. The expanded dim is num_layer which has the same position of seq dim.

Comment (qlzh727) [reply tensorflow/python/keras/layers/recurrent_v2.py]: The input init_h here is always [batch, num_units] (since keras only handles single layer), if the cudnnrnn ops expect the different shape of input for init_h based on the time/batch major, probably add a comment here since the init state has nothing to do with timestep (it is for timestep 0).

**Review by qlzh727 — DISMISSED** [review]

File: tensorflow/python/keras/layers/recurrent_v2.py (hunk 2 above)
Comment (houtoms) [reply tensorflow/python/keras/layers/recurrent_v2.py]: Yes, that makes sense. I will add some comments on this. Thx.

**Review by qlzh727 — APPROVED** [review]
